crates/bevy_animation/src/curves.rs

```diff
@@ -0,0 +1,702 @@
+    /// is returned instead.
+    #[inline]
+    pub fn duration(&self) -> Option<f32> {
+        match self {
+            VariableCurve::Translation(c) => c.duration(),
+            VariableCurve::Rotation(c) => c.duration(),
+            VariableCurve::Scale(c) => c.duration(),
+            VariableCurve::Weights(c) => c.duration(),
+        }
+    }
+}
+
+impl From<TranslationCurve> for VariableCurve {
+    fn from(curve: TranslationCurve) -> Self {
+        Self::Translation(curve)
+    }
+}
+
+impl From<RotationCurve> for VariableCurve {
+    fn from(curve: RotationCurve) -> Self {
+        Self::Rotation(curve)
+    }
+}
+
+impl From<ScaleCurve> for VariableCurve {
+    fn from(curve: ScaleCurve) -> Self {
+        Self::Scale(curve)
+    }
+}
+
+impl From<WeightsCurve> for VariableCurve {
+    fn from(curve: WeightsCurve) -> Self {
+        Self::Weights(curve)
+    }
+}
+
+//---------//
+// HELPERS //
+//---------//
+
+enum TwoIterators<A, B> {
+    Left(A),
+    Right(B),
+}
+
+impl<A, B, T> Iterator for TwoIterators<A, B>
+where
+    A: Iterator<Item = T>,
+    B: Iterator<Item = T>,
+{
+    type Item = T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self {
+            TwoIterators::Left(a) => a.next(),
+            TwoIterators::Right(b) => b.next(),
+        }
+    }
+}
+
+enum FourIterators<A, B, C, D> {
+    First(A),
+    Second(B),
+    Third(C),
+    Fourth(D),
+}
+
+impl<A, B, C, D, T> Iterator for FourIterators<A, B, C, D>
+where
+    A: Iterator<Item = T>,
+    B: Iterator<Item = T>,
+    C: Iterator<Item = T>,
+    D: Iterator<Item = T>,
+{
+    type Item = T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self {
+            FourIterators::First(a) => a.next(),
+            FourIterators::Second(b) => b.next(),
+            FourIterators::Third(c) => c.next(),
+            FourIterators::Fourth(d) => d.next(),
+        }
+    }
+}
+
+/// Helper function for cubic spline interpolation.
+fn cubic_spline_interpolation<T>(
+    value_start: T,
+    tangent_out_start: T,
+    tangent_in_end: T,
+    value_end: T,
+    lerp: f32,
+    step_duration: f32,
+) -> T
+where
+    T: VectorSpace,
+{
+    value_start * (2.0 * lerp.powi(3) - 3.0 * lerp.powi(2) + 1.0)
+        + tangent_out_start * (step_duration) * (lerp.powi(3) - 2.0 * lerp.powi(2) + lerp)
+        + value_end * (-2.0 * lerp.powi(3) + 3.0 * lerp.powi(2))
+        + tangent_in_end * step_duration * (lerp.powi(3) - lerp.powi(2))
+}
+
+fn cubic_spline_interpolate_slices<'a, T: VectorSpace>(
+    width: usize,
+    first: &'a [T],
+    second: &'a [T],
+    s: f32,
+    step_between: f32,
+) -> impl Iterator<Item = T> + 'a {
+    (0..width).map(move |idx| {
+        cubic_spline_interpolation(
+            first[idx + width],
+            first[idx + (width * 2)],
+            second[idx + width],
+            second[idx],
+            s,
+            step_between,
+        )
```

Comment:
Not really super deep into the code but comparing with the other occurrence of this function call we have

```rust
cubic_spline_interpolation(u[1], u[2], v[0], v[1], s, t1 - t0)
```

so in my mind I would think that we follow a similar pattern here. So to me it looks like the args for second got switched up and should be

```rust
cubic_spline_interpolation(
    first[idx + width],
    first[idx + (width * 2)],
    second[idx],
    second[idx + width],
    s,
    step_between,
)
```

since I think that `second[idx + width]` is the end value and `second[idx]` is the in tangent. But maybe I'm missing something, just making sure that this is intended